File: algorithms/199. Binary Tree Right Side View/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        right_values = []
        parents = [root]
        children = []
        while parents:
            right_values.append(parents[-1].val)
            for node in parents:
                if node.left:
                    children.append(node.left)
                if node.right:
                    children.append(node.right)
            parents, children = children, parents
            del children[:]
        return right_values

File: algorithms/199. Binary Tree Right Side View/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> rightmost;
        dfs(root, 0, rightmost);
        return rightmost;
    }

    void dfs(TreeNode * root, int curHeight, vector<int> & rightmost) {
        if ( root ) {
            if ( curHeight == rightmost.size() ) {
                rightmost.push_back(root->val);
            }
            dfs(root->right, curHeight + 1, rightmost);
            dfs(root->left, curHeight + 1, rightmost);
        }
    }
};
